### PR TITLE
SPARK-11406: Patch for a utf-8 decode issue that occurs when we get a…

### PR DESCRIPTION
**Pull request by boneill42**
… bad (non utf8) message through Kinesis.

### CODE REVIEW

**Comment by tdas** [comment]
ok to test.

**Comment by tdas** [comment]
Good catch. Could you add unit tests for this.

File: python/pyspark/streaming/kinesis.py [hunk 1]
```diff
@@ -26,6 +26,7 @@
 
 def utf8_decoder(s):
     """ Decode the unicode as UTF-8 """
+    s = unicode(s, errors='ignore')
     if s is None:
         return None
     return s.decode('utf-8')
```
Comment (tdas) [inline python/pyspark/streaming/kinesis.py]: isnt it better to change this line to 
`s.decode('utf-8',  errors='ignore')

**Comment by tdas** [comment]
this is ok to test

**Comment by tdas** [comment]
ok to test

**Comment by boneill42** [comment]
will do... 
Note: The case that triggered this is KPL aggregation.  I commented on SPARK-11198, and will explore a solution.

**Comment by boneill42** [comment]
I don't believe the new signature for decode is available in 2.6....

```
(spark)blu:streaming brianoneill$ python2.6
Python 2.6.9 (unknown, Aug 22 2015, 20:33:41)
>>> s = "foo"
>>> s.decode('utf-8', errors='ignore')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: decode() takes no keyword arguments

(spark)blu:streaming brianoneill$ python2.7
Python 2.7.10 (default, Jul 13 2015, 12:05:58)
>>> s = "foo"
>>> s.decode('utf-8', errors='ignore')
u'foo'
```

I've left the code as-is, and added a test.  Update coming...

**Comment by zsxwing** [comment]
I feel throwing an Error for invalid bytes does make sense. If people want to ignore such error, they can use a custom decoder.

**Comment by boneill42** [comment]
I'm inclined to agree as long as there is a way to catch that Exception and continue.  I'm not a python wizard, but it appeared as though the process died in worker.py, without giving the job a chance to catch the Exception.

**Comment by zsxwing** [comment]
> I'm inclined to agree as long as there is a way to catch that Exception and continue. I'm not a python wizard, but it appeared as though the process died in worker.py, without giving the job a chance to catch the Exception.

You can use the `decoder` parameter in the `createStream` method to set a custom decoder to catch the exception.

**Comment by boneill42** [comment]
I guess I'm just accustomed to explicit exceptions in Java, especially for something that might kill a job.  

But perhaps its sufficient to document this and let people know that everyone should implement a custom decoder if they want to protect against bad bytes in a record.

(feel free to close PR)

**Comment by zsxwing** [comment]
@boneill42 could you close this PR? We don't have permission to close it. It would be better if you can submit another PR to document this method. Thanks a lot!

**Comment by andrewor14** [comment]
@boneill42 can you close this issue?

**Comment by boneill42** [comment]
Yep, closed.

**Comment by obaidcuet** [comment]
Hi,

Just FYI. 
Hope it may be helpful for someone.

I had exactly same issue while reading twitter data from Kafka (collected using flume).
I have no idea where those non-UTF8 data coming from. The should be rejected by flume itself.
However, I do not need those special characters and used below to ignore them:

## create a decoder function with parameter "ignore"
def utf8_decoder_ignore_error(s):
    """ Decode the unicode as UTF-8 """
    if s is None:
        return None
    return s.decode('utf-8', "ignore")
	
## then use that decoder in createDirectStream as below
kvs = KafkaUtils.createDirectStream(ssc, [topic], {"metadata.broker.list": brokers}, keyDecoder=utf8_decoder_ignore_error, valueDecoder=utf8_decoder_ignore_error )	


-Obaid

**Pull request closed without merging**